### PR TITLE
read-replica-functionality-v1

### PR DESCRIPTION
**Pull request by gitmohit**
Adopting this sort of structure:

```
multiDatabase:
  tenantDbMap:
    shard1:
      url: "jdbc:mysql://localhost:3306/gst_stage_db
    readReplica1: # read only
      url: "jdbc:mysql://localhost:3306/gst_stage_db_read_1
    readReplica2: # read only
      url: "jdbc:mysql://localhost:3306/gst_stage_db_read_2
    readOnlyReplicaEnabled : true
    defaultReadReplicaTenant : "readReplica1"
```

### CODE REVIEW

File: sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/OrderIntegrationTestWithMultiTenancy.java [hunk 1]
```diff
@@ -58,7 +58,7 @@
 
     @Parameterized.Parameters
     public static List<ImmutablePair<String, String>> customerIds() {
-        return Lists.newArrayList(ImmutablePair.of("1", "shard1"),
+        return Lists.newArrayList(ImmutablePair.of("1", "shard1"));
```
Comment (gitmohit) [inline sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/OrderIntegrationTestWithMultiTenancy.java]: correct it

File: sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package in.cleartax.dropwizard.sharding.utils;
```
Comment (siddharth96) [inline sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java]: This shouldn't be in ```src``` but rather in ```test```.

File: sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * This sample application shows how to create and use a read-only database in a
+ * zip file. The database file is split into multiple smaller files, to speed up
+ * random-access. Splitting up the file is only needed if the database file is
+ * larger than a few megabytes.
+ */
+public class PrepareReadOnlyDB {
+
+    // TODO: We can get rid of hardcoding of filepath.
```
Comment (siddharth96) [inline sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java]: ??

**Review by siddharth96 — COMMENTED** [review]
Yet to review tests.

File: sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java (hunk 3 above)
Comment (gitmohit) [reply sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java]: Unnecessary comment removed it.

File: sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java (hunk 2 above)
Comment (gitmohit) [reply sharding-core/src/main/java/in/cleartax/dropwizard/sharding/utils/PrepareReadOnlyDB.java]: refactored it

File: sharding-core/pom.xml [hunk 4]
```diff
@@ -53,5 +53,10 @@
             <groupId>io.dropwizard.metrics</groupId>
             <artifactId>metrics-healthchecks</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <scope>compile</scope>
```
Comment (siddharth96) [inline sharding-core/pom.xml]: Shouldn't this be ```test```?

File: sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/OrderIntegrationTestWithReplica.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        return Lists.newArrayList(ImmutablePair.of(1, ImmutablePair.of("1", "readReplica")));
+    }
+
+    @BeforeClass
+    public static void setUp() {
+        client = TestSuiteWithReplicaEnabled.client;
+        host = String.format("http://localhost:%d/api", RULE.getLocalPort());
+    }
+
+    private void assertOrderPresentOnShard(final String expectedOnShard, final OrderDto orderDto) throws Throwable {
```
Comment (siddharth96) [inline sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/OrderIntegrationTestWithReplica.java]: Can you make it a Utility func?

File: sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/PrepareReadOnlyTestDB.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * This sample application shows how to create and use a read-only database in a
+ * zip file. The database file is split into multiple smaller files, to speed up
+ * random-access. Splitting up the file is only needed if the database file is
+ * larger than a few megabytes.
+ */
+public class PrepareReadOnlyTestDB {
+
+    public static void generateReplicaDB() throws Exception {
```
Comment (siddharth96) [inline sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/PrepareReadOnlyTestDB.java]: Remove path hardcoding.

File: sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/PrepareReadOnlyTestDB.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                "  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,\n" +
+                "  `name` varchar(255) DEFAULT NULL,\n" +
+                "  `order_id` int(11) DEFAULT NULL,\n" +
+                "  PRIMARY KEY (`id`)\n" +
+                ");");
+        stat.execute("TRUNCATE TABLE `orders`;\n" +
+                "TRUNCATE TABLE `order_items`;\n" +
+                "INSERT INTO `orders` (`id`, `order_ext_id`, `customer_id`, `amount`) VALUES ('1', '11111111-2222-3333-4444-aaaaaaaaaaa1', 1, 10000);\n" +
+                "INSERT INTO `order_items` (`id`, `name`, `order_id`) VALUES ('0', 'test', '10');\n" +
+                "INSERT INTO `order_items` (`id`, `name`, `order_id`) VALUES ('1', 'test', '10');");
```
Comment (siddharth96) [inline sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/PrepareReadOnlyTestDB.java]: Can you move this to a SQL file?

File: sharding-example/pom.xml [hunk 8]
```diff
@@ -60,5 +60,10 @@
             <artifactId>mockito-core</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <scope>compile</scope>
```
Comment (siddharth96) [inline sharding-example/pom.xml]: This should be ```test```, isn't it?

File: sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/PrepareReadOnlyTestDB.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            log.info("create the zip file...");
+            Backup.execute(readOnlyDbFileZip, readOnlyDbDirectoryPath, "", true);
+            String readOnlyDBUrl = "jdbc:h2:split:zip:" + readOnlyDbFileZip + "!/" + TEMP_DB_NAME + ";MODE=MySQL;DATABASE_TO_UPPER=false;IGNORECASE=TRUE;";
+
+            // delete the old database files
+            DeleteDbFiles.execute("split:"+readOnlyDbDirectoryPath, TEMP_DB_NAME, true);
+            return readOnlyDBUrl;
+        }
+        catch (Exception e) {
+            log.debug("Unable to create readOnly DB");
```
Comment (siddharth96) [inline sharding-example/src/test/java/in/cleartax/dropwizard/sharding/test/sampleapp/PrepareReadOnlyTestDB.java]: Log the exception for debugging later on. e.g.
```java
log.debug("Unable to create readOnly DB", e);
```

File: sharding-migrations/src/main/java/in/cleartax/dropwizard/sharding/migrations/DbGenerateDocsCommand.java [hunk 10]
```diff
@@ -7,7 +7,7 @@
 
 public class DbGenerateDocsCommand<T extends Configuration> extends AbstractLiquibaseCommand<T> {
     public DbGenerateDocsCommand(MultiTenantDatabaseConfiguration<T> strategy, Class<T> configurationClass, String migrationsFileName) {
-        super("generate-docs", "Generate documentation about the database state.", strategy, configurationClass, migrationsFileName);
+        super("generateReplicaDB-docs", "Generate documentation about the database state.", strategy, configurationClass, migrationsFileName);
```
Comment (siddharth96) [inline sharding-migrations/src/main/java/in/cleartax/dropwizard/sharding/migrations/DbGenerateDocsCommand.java]: Why?

File: sharding-migrations/src/test/java/in/cleartax/dropwizard/sharding/migrations/DbGenerateDocsCommandTest.java [hunk 11]
```diff
@@ -32,7 +32,7 @@ public void testHelpPage() throws Exception {
         final ByteArrayOutputStream baos = new ByteArrayOutputStream();
         createSubparser(generateDocsCommand).printHelp(new PrintWriter(new OutputStreamWriter(baos, UTF_8), true));
         assertThat(baos.toString(UTF_8)).isEqualTo(String.format(
-                "usage: db generate-docs [-h] [--migrations MIGRATIONS-FILE]%n" +
+                "usage: db generateReplicaDB-docs [-h] [--migrations MIGRATIONS-FILE]%n" +
```
Comment (siddharth96) [inline sharding-migrations/src/test/java/in/cleartax/dropwizard/sharding/migrations/DbGenerateDocsCommandTest.java]: Same here.

File: sharding-migrations/src/test/java/in/cleartax/dropwizard/sharding/migrations/DbCommandTest.java [hunk 12]
```diff
@@ -38,13 +38,13 @@ public void testPrintHelp() throws Exception {
...
                 "usage: db db [-h]%n" +
-                        "          {calculate-checksum,clear-checksums,drop-all,dump,fast-forward,generate-docs,locks,migrate,prepare-rollback,rollback,status,tag,test}%n" +
+                        "          {calculate-checksum,clear-checksums,drop-all,dump,fast-forward,generateReplicaDB-docs,locks,migrate,prepare-rollback,rollback,status,tag,test}%n" +
                         "          ...%n" +
                         "%n" +
                         "Run database migration tasks%n" +
                         "%n" +
                         "positional arguments:%n" +
-                        "  {calculate-checksum,clear-checksums,drop-all,dump,fast-forward,generate-docs,locks,migrate,prepare-rollback,rollback,status,tag,test}%n" +
+                        "  {calculate-checksum,clear-checksums,drop-all,dump,fast-forward,generateReplicaDB-docs,locks,migrate,prepare-rollback,rollback,status,tag,test}%n" +
```
Comment (siddharth96) [inline sharding-migrations/src/test/java/in/cleartax/dropwizard/sharding/migrations/DbCommandTest.java]: And here.

**Review by siddharth96 — APPROVED** [review]

**Pull request merged**